src/style-spec/function/expression.js

```diff
@@ -0,0 +1,304 @@
+    Color: {}
+};
+
+Type.Array = {
+    [Type.Any]: {},
+    [Type.Number]: {},
+    [Type.String]: {},
+    [Type.Boolean]: {},
+    [Type.Color]: {}
+};
+
+module.exports = compileExpression;
+
+function compileExpression(expr) {
+    const compiled = compile(expr);
+    const fun = new Function('mapProperties', 'feature', `
+var props = (feature && feature.properties || {});
+return (${compiled.compiledExpression})
+`);
+    fun.isFeatureConstant = compiled.isFeatureConstant;
+    fun.isZoomConsant = compiled.isZoomConstant;
+    return fun;
+}
+
+const functions = {
+    'ln2': {
+        input: [],
+        output: Type.Number,
+    },
+    'pi': {
+        input: [],
+        output: Type.Number,
+    },
+    'e': {
+        input: [],
+        output: Type.Number,
+    },
+    'zoom': {
+        input: [],
+        output: Type.Number,
+    },
+    'data': {
+        input: [Type.String],
+        output: Type.Any,
+    },
+    'has': {
+        input: [Type.String],
+        output: Type.Boolean
+    },
+    '+': {
+        input: [Type.Number, Type.Number],
+        output: Type.Number
+    },
+    '*': {
+        input: [Type.Number, Type.Number],
+        output: Type.Number
+    },
+    '-': {
+        input: [Type.Number, Type.Number],
+        output: Type.Number
+    },
+    '/': {
+        input: [Type.Number, Type.Number],
+        output: Type.Number
+    },
+    '^': {
+        input: [Type.Number, Type.Number],
+        output: Type.Number
+    },
+    '%': {
+        input: [Type.Number, Type.Number],
+        output: Type.Number
+    },
+    'log10': {
+        input: [Type.Number],
+        output: Type.Number,
+        math: true
+    },
+    'ln': {
+        input: [Type.Number],
+        output: Type.Number,
+        math: true
+    },
+    'log2': {
+        input: [Type.Number],
+        output: Type.Number,
+        math: true
+    },
+    'sin': {
+        input: [Type.Number],
+        output: Type.Number,
+        math: true
+    },
+    'cos': {
+        input: [Type.Number],
+        output: Type.Number,
+        math: true
+    },
+    'tan': {
+        input: [Type.Number],
+        output: Type.Number,
+        math: true
+    },
+    'asin': {
+        input: [Type.Number],
+        output: Type.Number,
+        math: true
+    },
+    'acos': {
+        input: [Type.Number],
+        output: Type.Number,
+        math: true
+    },
+    'atan': {
+        input: [Type.Number],
+        output: Type.Number,
+        math: true
+    },
+    'ceil': {
+        input: [Type.Number],
+        output: Type.Number,
+        math: true
+    },
+    'floor': {
+        input: [Type.Number],
+        output: Type.Number,
+        math: true
+    },
+    'round': {
+        input: [Type.Number],
+        output: Type.Number,
+        math: true
+    },
+    'abs': {
+        input: [Type.Number],
+        output: Type.Number,
+        math: true
+    },
+    'min': {
+        input: [Type.Array[Type.Number]],
```

Comment:
Currently, `Type.Array[t]` really just means 'n arguments of type t'; given that we'll probably end up needing to use that for actual array values (like `["array", e0, e1, e2...]`), this should probably be changed to something like `Type.Nargs[t]`.